wcfsetup/install/files/lib/system/background/job/UnfurlUrlBackgroundJob.class.php

```diff
@@ -0,0 +1,275 @@
+
+            if (empty(StringUtil::trim($unfurlResponse->getTitle()))) {
+                $this->save(UnfurlUrl::STATUS_REJECTED);
+
+                return;
+            }
+
+            $title = StringUtil::truncate(StringUtil::trim($unfurlResponse->getTitle()), 255);
+            $description = "";
+            if ($unfurlResponse->getDescription()) {
+                $description = StringUtil::truncate(StringUtil::trim($unfurlResponse->getDescription()), 160);
+            }
+
+            $imageData = [];
+            $imageID = null;
+            if ($unfurlResponse->getImageUrl()) {
+                $imageUrl = $unfurlResponse->getImageUrl();
+
+                if (
+                    \strpos($imageUrl, '\\') === false
+                    && \strpos($imageUrl, "'") === false
+                    && Url::is($imageUrl)
+                ) {
+                    $imageID = self::getImageIdByUrl($imageUrl);
+
+                    if ($imageID === null) {
+                        $imageData = $this->getImageData($unfurlResponse);
+                    }
+                }
+            }
+
+            $this->save(
+                UnfurlUrl::STATUS_SUCCESSFUL,
+                $title,
+                $description,
+                $imageID,
+                $imageData
+            );
+        } catch (UrlInaccessible | ParsingFailed $e) {
+            if (\ENABLE_DEBUG_MODE) {
+                \wcf\functions\exception\logThrowable($e);
+            }
+
+            $this->save(UnfurlUrl::STATUS_REJECTED);
+        }
+    }
+
+    private function getImageData(UnfurlResponse $unfurlResponse): array
+    {
+        $imageSaveData = [];
+
+        if (empty($unfurlResponse->getImageUrl()) || !Url::is($unfurlResponse->getImageUrl())) {
+            throw new \InvalidArgumentException("The given response does not have an image.");
+        }
+
+        try {
+            $imageResponse = $unfurlResponse->getImage();
+            $image = $this->downloadImage($imageResponse);
+            $imageData = \getimagesizefromstring($image);
+
+            if ($imageData !== false) {
+                if ($this->validateImage($imageData)) {
+                    $imageSaveData['imageUrl'] = $unfurlResponse->getImageUrl();
+                    $imageSaveData['imageUrlHash'] = \sha1($unfurlResponse->getImageUrl());
+                    $imageSaveData['width'] = $imageData[0];
+                    $imageSaveData['height'] = $imageData[1];
+                    if (!(MODULE_IMAGE_PROXY || IMAGE_ALLOW_EXTERNAL_SOURCE)) {
+                        $this->saveImage($imageData, $image, $imageSaveData['imageUrlHash']);
+                        $imageSaveData['imageExtension'] = $this->getImageExtension($imageData);
+                        $imageSaveData['isStored'] = 1;
+                    }
+                }
+            }
+        } catch (UrlInaccessible | DownloadFailed $e) {
+            $imageSaveData = [];
+        }
+
+        return $imageSaveData;
```

Comment:
```suggestion
        $imageSaveData = [];

        if (empty($unfurlResponse->getImageUrl()) || !Url::is($unfurlResponse->getImageUrl())) {
            throw new \InvalidArgumentException("The given response does not have an image.");
        }

        try {
            $imageResponse = $unfurlResponse->getImage();
            $image = $this->downloadImage($imageResponse);
            $imageData = \getimagesizefromstring($image);

            if ($imageData === false) {
                return [];
            }
                if (!$this->validateImage($imageData)) {
                    return [];
                }
                $imageSaveData = [
                 'imageUrl' => $unfurlResponse->getImageUrl(),
                    'imageUrlHash'] => \sha1($unfurlResponse->getImageUrl()),
                    // …
                    ];
                    if (!(MODULE_IMAGE_PROXY || IMAGE_ALLOW_EXTERNAL_SOURCE)) {
                        $this->saveImage($imageData, $image, $imageSaveData['imageUrlHash']);
                        $imageSaveData['imageExtension'] = $this->getImageExtension($imageData);
                        $imageSaveData['isStored'] = 1;
                    }
                    return $imageSaveData;
        } catch (UrlInaccessible | DownloadFailed $e) {
            return [];
        }
```

Ignore the formatting, you get the point: Reduce the nesting level here and use early returns.

Reply:
I don't think, that this is more readable, but I will apply it…